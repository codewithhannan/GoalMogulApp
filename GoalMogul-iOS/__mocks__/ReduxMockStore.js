// __mocks__/redux-mock-store.js
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [ thunk ];
const mockStore = configureMockStore(middlewares);

export default mockStore;