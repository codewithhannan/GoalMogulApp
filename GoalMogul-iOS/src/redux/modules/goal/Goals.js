/**
 * This reducer is the source of truth for Goals related Components based on the design
 * https://docs.google.com/document/d/1JxjCLZGsp5x_Sy0R56y05aUgAUALtuSVAgJIway9JTg/edit?usp=sharing
 */

import _ from 'lodash';

// Constants
import {
    PROFILE_FETCH_TAB_DONE,
    PROFILE_REFRESH_TAB_DONE,
    PROFILE_GOAL_DELETE_SUCCESS
} from '../../../reducers/Profile';

import {
    PROFILE_CLOSE_PROFILE
} from '../../../actions/types';

/**
 * List of const to add 
 * 
 * Create Goal
 * GOAL_CREATE_EDIT_SUCCESS
 * 
 * Goal Detail related
 * GOAL_DETAIL_MARK_NEED_AS_COMPLETE_SUCCESS
 * GOAL_DETAIL_MARK_AS_COMPLETE_SUCCESS,
 * GOAL_DETAIL_SHARE_TO_MASTERMIND_SUCCESS,
 * GOAL_DETAIL_MARK_STEP_AS_COMPLETE_SUCCESS,
 * GOAL_DETAIL_MARK_NEED_AS_COMPLETE_SUCCESS
 * 
 * Profile related (done)
 * PROFILE_FETCH_TAB_DONE (done)
 * PROFILE_REFRESH_TAB_DONE (done)
 * PROFILE_GOAL_DELETE_SUCCESS (done)
 * PROFILE_CLOSE_PROFILE (done)
 * 
 * Comment related
 * The ones that need to increase / decrease comment count
 * 
 * Home related (Goal Feed)
 * 
 */

const DEBUG_KEY = '[ Reducer Goals ]';

// Sample goal object in the map
const INITIAL_GOAL = {
    goal: {},
    pageId: {
        refreshing: false
    },
    reference: [],
};

const INITIAL_GOAL_PAGE = {
    refreshing: false,
    // Potential navigation state and etc. First focus on integration with Profile
};

const INITIAL_STATE = {

};

export default (state = INITIAL_STATE, action) => {
    switch (action.type) {
        /* Profile related */
        case PROFILE_REFRESH_TAB_DONE:
        case PROFILE_FETCH_TAB_DONE: {
            const { pageId, data, type } = action.payload;
            let newState = _.cloneDeep(state);
            if (type !== 'goals') return newState;
            if (!data || _.isEmpty(data)) return newState;

            data.forEach(goal => {
                const goalId = goal._id;
                // Update goal
                if (!_.has(newState, goalId)) {
                    newState = _.set(newState, `${goalId}.goal`, goal);
                }

                const oldReference = _.get(newState, `${goalId}.reference`);
                const hasPageReference = (oldReference !== undefined && oldReference.some(r => r === pageId));
                // Update reference
                let newReference = [pageId];
                if (!hasPageReference) {
                    newReference = newReference.concat(oldReference);
                }
                newState = _.set(newState, `${goalId}.reference`, newReference);
            });

            return newState;
        }

        case PROFILE_GOAL_DELETE_SUCCESS: {
            const { pageId, goalId } = action.payload;
            let newState = _.cloneDeep(state);
            if (!_.has(newState, goalId)) return newState; // no reference to remove

            let goalToUpdate = _.get(newState, `${goalId}`);

            // Update reference
            const oldReference = _.get(goalToUpdate, 'reference');
            let newReference = oldReference;
            if (oldReference && oldReference.some(r => r === pageId)) {
                newReference = newReference.filter(r => r !== pageId);
            }

            // Remove pageId reference object
            goalToUpdate = _.omit(goalToUpdate, `${pageId}`);

            // Remove this goal if it's no longer referenced
            if (!newReference || _.isEmpty(newReference)) {
                newState = _.omit(newState, `${goalId}`);
                return newState;
            }

            // Update the goal by goalId
            goalToUpdate = _.set(goalToUpdate, 'reference', newReference);
            newState = _.set(newState, `${goalId}`, goalToUpdate);
            return newState;
        }

        case PROFILE_CLOSE_PROFILE: {
            const { goalList, pageId } = action.payload;
            let newState = _.cloneDeep(state);
            if (!goalList || _.isEmpty(goalList)) return newState;

            goalList.forEach(goalId => {
                // Check if goalId in the Goals
                if (!_.has(newState, goalId)) {
                    return;
                }

                let goalToUpdate = _.get(newState, `${goalId}`);

                // Remove pageId reference object
                goalToUpdate = _.omit(goalToUpdate, `${pageId}`);

                const oldReference = _.get(newState, `${goalId}.reference`);
                const hasPageReference = (oldReference !== undefined && oldReference.some(r => r === pageId));
                // Remove reference
                let newReference = oldReference;
                if (hasPageReference) {
                    newReference = oldReference.filter(r => r !== pageId);
                }

                // Remove this goal if it's no longer referenced
                if (!newReference || _.isEmpty(newReference)) {
                    newState = _.omit(newState, `${goalId}`);
                    return;
                }
                goalToUpdate = _.set(goalToUpdate, 'reference', newReference);
                newState = _.set(newState, `${goalId}`, goalToUpdate);
            });

            return newState;
        }

        default: 
            return { ...state };
    }
};
