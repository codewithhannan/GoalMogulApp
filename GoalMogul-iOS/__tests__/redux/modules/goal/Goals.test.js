import _ from 'lodash';
import Goals from '../../../../src/redux/modules/goal/Goals';

import {
    GOAL_DETAIL_OPEN,
    GOAL_DETAIL_FETCH,
    GOAL_DETAIL_FETCH_DONE,
    GOAL_DETAIL_FETCH_ERROR,
    GOAL_DETAIL_UPDATE,
    GOAL_DETAIL_UPDATE_DONE,
    GOAL_DETAIL_CLOSE,
    GOAL_DETAIL_MARK_NEED_AS_COMPLETE_SUCCESS,
    GOAL_DETAIL_MARK_STEP_AS_COMPLETE_SUCCESS,
    GOAL_DETAIL_SHARE_TO_MASTERMIND_SUCCESS,
    GOAL_DETAIL_MARK_AS_COMPLETE_SUCCESS,
    GOAL_DETAIL_SWITCH_TAB_V2
} from '../../../../src/reducers/GoalDetailReducers';
import { mockAGoal, DEFAULT_GOAL } from '../../../../__mocks__/mockGoals';

export const initialState = {
    goal: {},
    reference: [],
};

describe('Test page reference update', () => {
    let sharedGoalStore = _.cloneDeep(initialState);
    it('Same goal is opened twice', () => {
        const action = {
            type: GOAL_DETAIL_OPEN,
            payload: {
                goal: DEFAULT_GOAL,
                pageId: 1,
                goalId: _.get(DEFAULT_GOAL, "_id")
            }
        };
        const secondAction = {
            type: GOAL_DETAIL_OPEN,
            payload: {
                goal: DEFAULT_GOAL,
                pageId: 2,
                goalId: _.get(DEFAULT_GOAL, "_id")
            }
        };

        sharedGoalStore = Goals(sharedGoalStore, action);
        expect(sharedGoalStore).toMatchSnapshot();

        sharedGoalStore = Goals(sharedGoalStore, secondAction);
        expect(sharedGoalStore).toMatchSnapshot();
    });

    it('A goal page is closed', () => {
        const closeGoalAction = {
            type: GOAL_DETAIL_CLOSE,
            payload: {
                pageId: 1,
                goalId: _.get(DEFAULT_GOAL, "_id")
            }
        };
        sharedGoalStore = Goals(sharedGoalStore, closeGoalAction);
        expect(sharedGoalStore).toMatchSnapshot();
    });
});

describe('Test specific goal update', () => {

});
