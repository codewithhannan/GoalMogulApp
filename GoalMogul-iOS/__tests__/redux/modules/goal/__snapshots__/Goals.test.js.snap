// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A goal page is closed 1`] = `
Object {
  "5b502211e500e3001afd1e20": Object {
    "2": Object {
      "loading": false,
      "navigationStateV2": Object {
        "focusRef": undefined,
        "focusType": undefined,
        "index": 0,
        "routes": Array [
          Object {
            "key": "centralTab",
            "title": "CentralTab",
          },
          Object {
            "key": "focusTab",
            "title": "FocusTab",
          },
        ],
      },
      "refreshing": false,
      "updating": false,
    },
    "goal": Object {
      "__v": 0,
      "_id": "5b502211e500e3001afd1e20",
      "category": "General",
      "created": "2018-07-19T05:30:57.531Z",
      "details": Object {
        "tags": Array [],
        "text": "This is detail",
      },
      "feedInfo": Object {
        "_id": "5b502211e500e3001afd1e18",
        "publishDate": "2018-07-19T05:30:57.531Z",
      },
      "lastUpdated": "2018-07-19T05:30:57.531Z",
      "needs": Array [
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "introduction to someone from the Bill and Melinda Gates Foundation",
          "isCompleted": false,
          "order": 0,
        },
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "Get in contact with Nuclear experts",
          "isCompleted": false,
          "order": 1,
        },
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "Legal & Safety experts who have worked with the United States",
          "isCompleted": false,
          "order": 2,
        },
      ],
      "owner": Object {
        "_id": "5b17781ebec96d001a409960",
        "name": "jia zeng",
        "profile": Object {
          "elevatorPitch": "This is my elevatorPitch",
          "occupation": "Software Engineer",
          "pointsEarned": 10,
          "views": 0,
        },
      },
      "priority": 3,
      "privacy": "friends",
      "steps": Array [
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "This is my first step to complete the goal",
          "isCompleted": false,
          "order": 0,
        },
      ],
      "title": "Establish a LMFBR near Westport, Connecticut by 2020",
    },
    "reference": Array [
      2,
    ],
  },
  "goal": Object {},
  "reference": Array [],
}
`;

exports[`Same goal is opened twice 1`] = `
Object {
  "5b502211e500e3001afd1e20": Object {
    "1": Object {
      "loading": false,
      "navigationStateV2": Object {
        "focusRef": undefined,
        "focusType": undefined,
        "index": 0,
        "routes": Array [
          Object {
            "key": "centralTab",
            "title": "CentralTab",
          },
          Object {
            "key": "focusTab",
            "title": "FocusTab",
          },
        ],
      },
      "refreshing": false,
      "updating": false,
    },
    "goal": Object {
      "__v": 0,
      "_id": "5b502211e500e3001afd1e20",
      "category": "General",
      "created": "2018-07-19T05:30:57.531Z",
      "details": Object {
        "tags": Array [],
        "text": "This is detail",
      },
      "feedInfo": Object {
        "_id": "5b502211e500e3001afd1e18",
        "publishDate": "2018-07-19T05:30:57.531Z",
      },
      "lastUpdated": "2018-07-19T05:30:57.531Z",
      "needs": Array [
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "introduction to someone from the Bill and Melinda Gates Foundation",
          "isCompleted": false,
          "order": 0,
        },
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "Get in contact with Nuclear experts",
          "isCompleted": false,
          "order": 1,
        },
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "Legal & Safety experts who have worked with the United States",
          "isCompleted": false,
          "order": 2,
        },
      ],
      "owner": Object {
        "_id": "5b17781ebec96d001a409960",
        "name": "jia zeng",
        "profile": Object {
          "elevatorPitch": "This is my elevatorPitch",
          "occupation": "Software Engineer",
          "pointsEarned": 10,
          "views": 0,
        },
      },
      "priority": 3,
      "privacy": "friends",
      "steps": Array [
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "This is my first step to complete the goal",
          "isCompleted": false,
          "order": 0,
        },
      ],
      "title": "Establish a LMFBR near Westport, Connecticut by 2020",
    },
    "reference": Array [
      1,
    ],
  },
  "goal": Object {},
  "reference": Array [],
}
`;

exports[`Same goal is opened twice 2`] = `
Object {
  "5b502211e500e3001afd1e20": Object {
    "1": Object {
      "loading": false,
      "navigationStateV2": Object {
        "focusRef": undefined,
        "focusType": undefined,
        "index": 0,
        "routes": Array [
          Object {
            "key": "centralTab",
            "title": "CentralTab",
          },
          Object {
            "key": "focusTab",
            "title": "FocusTab",
          },
        ],
      },
      "refreshing": false,
      "updating": false,
    },
    "2": Object {
      "loading": false,
      "navigationStateV2": Object {
        "focusRef": undefined,
        "focusType": undefined,
        "index": 0,
        "routes": Array [
          Object {
            "key": "centralTab",
            "title": "CentralTab",
          },
          Object {
            "key": "focusTab",
            "title": "FocusTab",
          },
        ],
      },
      "refreshing": false,
      "updating": false,
    },
    "goal": Object {
      "__v": 0,
      "_id": "5b502211e500e3001afd1e20",
      "category": "General",
      "created": "2018-07-19T05:30:57.531Z",
      "details": Object {
        "tags": Array [],
        "text": "This is detail",
      },
      "feedInfo": Object {
        "_id": "5b502211e500e3001afd1e18",
        "publishDate": "2018-07-19T05:30:57.531Z",
      },
      "lastUpdated": "2018-07-19T05:30:57.531Z",
      "needs": Array [
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "introduction to someone from the Bill and Melinda Gates Foundation",
          "isCompleted": false,
          "order": 0,
        },
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "Get in contact with Nuclear experts",
          "isCompleted": false,
          "order": 1,
        },
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "Legal & Safety experts who have worked with the United States",
          "isCompleted": false,
          "order": 2,
        },
      ],
      "owner": Object {
        "_id": "5b17781ebec96d001a409960",
        "name": "jia zeng",
        "profile": Object {
          "elevatorPitch": "This is my elevatorPitch",
          "occupation": "Software Engineer",
          "pointsEarned": 10,
          "views": 0,
        },
      },
      "priority": 3,
      "privacy": "friends",
      "steps": Array [
        Object {
          "created": "2018-07-19T05:30:57.531Z",
          "description": "This is my first step to complete the goal",
          "isCompleted": false,
          "order": 0,
        },
      ],
      "title": "Establish a LMFBR near Westport, Connecticut by 2020",
    },
    "reference": Array [
      2,
      1,
    ],
  },
  "goal": Object {},
  "reference": Array [],
}
`;
